### PR TITLE
Reduce size of the Action enum using Box

### PR DESCRIPTION
**Pull request by saketh-are**
**TL;DR**: The `Action` enum has several variants of differing sizes. By Box-ing the larger variants, this PR significantly reduces the memory usage for almost all of `Action`'s variants. Only the largest, `Action::SignedDelegate`, sees a slight increase in total memory usage from 224 Bytes to 248 Bytes.

### Detailed Outcome

| Variant | Size (Bytes) | Boxed in this PR | Memory Usage (Before) | Memory Usage (After) |
| ------- | ------------ | ---------------- | ---------------------------|---------------------------|
CreateAccount | 0 | | 224 | 32 |
DeployContract | 24 | | 224 | 32 |
FunctionCall | 80 | Yes | 224 | 112 |
Transfer | 16 | | 224 | 32 |
Stake | 96 | Yes | 224 | 128 |
AddKey | 176 | Yes | 224 | 208 |
DeleteKey | 65 | Yes | 224 | 97 |
DeleteAccount | 16 | | 224 | 32 |
SignedDelegate | 216 | Yes | 224 | 248 |

One specific benefit of this change is reducing decoding overhead for network messages. For example, an action of type `Action::CreateAccount` occupies a single byte in serialized network format. When a message's single byte can deserialize back to 224 bytes in memory, it places a higher and less easily predictable resource demand on the node. This PR makes progress on that front by reducing the deserialized size for `Action::CreateAccount` to 32 bytes.

### How does it work?

The size of an enum is statically determined by the size of its largest variant plus the size of its discriminant (some bits telling us which variant is stored). In this case, `Action::SignedDelegate` is the largest variant; it  stores a struct `SignedDelegateAction` with size 216 bytes. After adding a discriminant and some padding bits for memory alignment, the `Action` enum reaches its total size of 224 Bytes.

`Box<T>` is a pointer type allocating memory on the heap to store data of type `T`. By having the `Action::SignedDelegate` variant store a `Box<SignedDelegateAction>`, we reduce its direct size from 224 Bytes to 8 Bytes, the size of a Box pointer on a 64-bit machine. Indirectly, it allocates an additional 224 Bytes to store the value of the `SignedDelegateAction` struct.

By Box-ing the large `Action` variants and reducing their direct size to 8 bytes, we leave `Action::DeployContract` as the variant with the largest direct size at 24 Bytes. Adding the discriminant and padding leaves `Action` with a new total size of 32 Bytes.

#### Performance Considerations
Generally, reducing the size of a datatype will also reduce the processing time associated with writes or reads on it.

However, use of Box comes with some small performance penalties. Box's memory indirection means that reading or writing a Boxed value involves 1 additional memory lookup. It also means that Box-ed variants stored in a Vec<Action> no longer have all their data stored contiguously in the Vec, weakening benefits of memory caching. If a variant is small enough already, a small reduction in size may not bring enough benefit to offset these other penalties.

Without good benchmarks on node performance to guide our decision-making, this PR makes the choice to Box the very large variants (>64 Bytes) and leave smaller ones untouched.

#### Limited Remaining Upside

The size of an enum is determined by the size of its largest variant plus the size of its discriminant (some bits telling us which variant is stored). On a 64-bit machine, Box occupies 8 bytes. After adding a discriminant and padding for memory alignment, the minimum possible size for an Enum with a Box-ed variant is 16 Bytes. We've already reduced Action to 32 Bytes in this PR by addressing the very large variants.

### CODE REVIEW

**Review by wacban — APPROVED** [review]
🎁

**Comment by jakmeier** [comment]
Wow, thanks a lot for the detailed motivation and explanations! It makes it clear that there is potentially a lot of memory overhead for actions in the current format.

> Without good benchmarks on node performance to guide our decision-making, this PR makes the choice to Box the very large variants (>64 Bytes) and leave smaller ones untouched.

Curious: How did we identify this as a bottleneck (memory or CPU) without some sort of benchmark?

Follow-up question: Is there maybe some quick way we could run a simplified benchmark for this? At least to check that additional allocations while decoding are not too big of an issue for what we can tolerate. I'm also curious if it affects the performance of hashing the borsh serialized format, which I believe we do in several places in the code.

File: runtime/runtime/src/config.rs [hunk 1]
```diff
@@ -107,7 +106,7 @@ pub fn total_send_fees(
                 transfer_send_fee(config, sender_is_receiver, is_receiver_implicit)
             }
             Stake(_) => config.fee(ActionCosts::stake).send_fee(sender_is_receiver),
-            AddKey(AddKeyAction { access_key, .. }) => match &access_key.permission {
+            AddKey(add_key_action) => match add_key_action.access_key.permission.clone() {
```
Comment (nagisa) [inline runtime/runtime/src/config.rs:109]: Is cloning here really necessary? Seems like handling this by-reference should be possible.

File: runtime/runtime/src/config.rs [hunk 2]
```diff
@@ -210,7 +210,7 @@ pub fn exec_fee(
             transfer_exec_fee(config, is_receiver_implicit)
         }
         Stake(_) => config.fee(ActionCosts::stake).exec_fee(),
-        AddKey(AddKeyAction { access_key, .. }) => match &access_key.permission {
+        AddKey(add_key_action) => match add_key_action.access_key.permission.clone() {
```
Comment (nagisa) [inline runtime/runtime/src/config.rs:213]: Same here.

**Review by nagisa — APPROVED** [review]
